api/call.go

```diff
@@ -154,48 +158,90 @@ func (ar *Runner) callHandler() func(http.ResponseWriter, *http.Request) {
+			processError(err, err.Error(), &contractAnswer, insLog, traceID)
 			return
 		}
 
-		if params.LogLevel != nil {
-			logLevelNumber, err := insolar.ParseLevel(*params.LogLevel)
+		contractAnswer.JSONRPC = contractRequest.JSONRPC
+		contractAnswer.ID = contractRequest.ID
+
+		digest := req.Header.Get(requester.Digest)
+		richSignature := req.Header.Get(requester.Signature)
+
+		signature, err := validateRequestHeaders(digest, richSignature, rawBody)
+		if err != nil {
+			processError(err, err.Error(), &contractAnswer, insLog, traceID)
+			return
+		}
+
+		if len(contractRequest.LogLevel) > 0 {
+			logLevelNumber, err := insolar.ParseLevel(contractRequest.LogLevel)
 			if err != nil {
-				processError(err, "Can't parse logLevel", &resp, insLog)
+				processError(err, "Can't parse logLevel", &contractAnswer, insLog, traceID)
 				return
 			}
 			ctx = inslogger.WithLoggerLevel(ctx, logLevelNumber)
 		}
 
-		err = ar.checkSeed(params.Seed)
+		err = ar.checkSeed(contractRequest.Params.Seed)
 		if err != nil {
-			processError(err, "Can't checkSeed", &resp, insLog)
+			processError(err, err.Error(), &contractAnswer, insLog, traceID)
 			return
 		}
 
+		// pulse, err := ar.PulseAccessor.Latest(ctx)
+		// if err != nil {
+		// 	processError(err, "Can't get last pulse", &contractAnswer, insLog, traceID)
+		// 	return
+		// }
+
 		var result interface{}
 		ch := make(chan interface{}, 1)
 		go func() {
-			result, err = ar.makeCall(ctx, params)
+			// result, err = ar.makeCall(ctx, contractRequest, rawBody, signature, pulse.PulseTimestamp)
+			result, err = ar.makeCall(ctx, contractRequest, rawBody, signature, 0)
 			ch <- nil
 		}()
 		select {
 
 		case <-ch:
 			if err != nil {
-				processError(err, "Can't makeCall", &resp, insLog)
+				processError(err, err.Error(), &contractAnswer, insLog, traceID)
 				return
 			}
-			resp.Result = result
+			contractResult := &requester.Result{ContractResult: result, TraceID: traceID}
+			contractAnswer.Result = contractResult
+			return
 
 		case <-time.After(time.Duration(ar.cfg.Timeout) * time.Second):
-			resp.Error = "Messagebus timeout exceeded"
+			errResponse := &requester.Error{Message: "Messagebus timeout exceeded", Code: TimeoutError, TraceID: traceID}
+			contractAnswer.Error = errResponse
 			return
+		}
+	}
+}
 
+func validateRequestHeaders(digest string, richSignature string, body []byte) (string, error) {
+	// Digest = "SHA-256=<hashString>"
+	// Signature = "keyId="member-pub-key", algorithm="ecdsa", headers="digest", signature=<signatureString>"
+	if len(digest) == 0 || len(richSignature) == 0 || len(body) == 0 {
+		return "", errors.Errorf("Invalid input data length digest: %d, signature: %d, body: %d", len(digest), len(richSignature), len(body))
+	}
+	h := sha256.New()
+	_, err := h.Write(body)
+	if err != nil {
+		return "", errors.Wrap(err, "Cant get hash")
+	}
+	sha := base64.StdEncoding.EncodeToString(h.Sum(nil))
+	if sha == digest[strings.IndexByte(digest, '=')+1:] {
```

Comment:
where would be panic if the digest is in not expected format: https://play.golang.org/p/pvjtsuF3mGM

also this: https://play.golang.org/p/lHGEJhCbBec

Reply:
Here we took the last part of a string
Digest=SHA-256=foo
or 
Digest=SHA-256=bar=
But anyway, need to validate the digest more precise